keps/sig-network/2594-multiple-cluster-cidrs/README.md

```diff
@@ -0,0 +1,980 @@
+update the node object with the podCIDRs.
+
+##### Node Updated
+
+Check that its Pod CIDR(s) match internal allocation.
+
+-   If node.spec.PodCIDRs is already filled up, honor that allocation and mark
+    those ranges as allocated.
+-   If the node.spec.PodCIDRs is filled with a CIDR not from any
+    `ClusterCIDRConfig`, raise a K8sEvent.
+-   If the ranges are already marked as allocated for some other node, raise
+    another error event (there isn’t an obvious reconciliation step the
+    controller can take unilaterally).
+
+##### Node Deleted
+
+Release said Node’s allocation from the internal data-structure.
+
+If this Node is the last one using a particular `ClusterCIDRConfig` that has
+been slated for deletion, trigger the deletion flow again (so that the finalizer
+is removed and internal data structures are cleaned up).
+
+##### ClusterCIDRConfig Added
+
+Install a finalizer on the `ClusterCIDRConfig` called "networking.kubernetes.io/cluster-cidr-config-finalizer".
+
+Update internal representation of CIDRs to include the new range. Every failed
+Node Allocation is stored in a queue, that will be tried again with the new
+range by the reconciliation loop.
+
+##### ClusterCIDRConfig Updated
+
+_`CIDR`, and `PerNodeMaskSize` are immutable_
+
+Update any internal state as required.
+
+If the `NodeSelector` changed to include a node that is currently awaiting a
+PodCIDR allocation, it will be allocated by the reconiliation loop. 
+
+If the `NodeSelector` changes to unselect a Node currently using that range,
+there is no change to the Node's allocation.
+
+##### ClusterCIDRConfig Deleted
+
+1.  Update internal data structures to mark the range as terminating (so new
+    nodes won't be added to it)
+1.  Change the status of the ClusterCIDRConfig with the following conditions:
+    `[{"active": false}, {"terminating": true}]`.
+1.  Search the internal representation of the CIDR range to see if any Nodes are
+    using the range. If they are, update the `"terminating"` condition with the
+    `reason: "cidr-in-use"`.
+1.  If there are no nodes using the range, remove the finalizer and cleanup all
+    internal state.
+
+### kube-controller-manager
+
+The flag `--cidr-allocator-type` will be ammended to include a new type
+"ClusterCIDRConfigAllocator".
+
+The list of current valid types is
+[here](https://github.com/kubernetes/kubernetes/blob/1ff18a9c43f59ffed3b2d266b31e0d696d04eaff/pkg/controller/nodeipam/ipam/cidr_allocator.go#L38).
+
+### Test Plan
+
+#### Unit Tests and Benchmarks
+
+-   Ensure that the controller scales to ~5,000 nodes -- memory usage and reasonable
+    allocation times
+
+#### Integration Tests
+
+-   Verify finalizers and statuses are persisted appropriately
+-   Test watchers
+
+#### End-to-End Tests
+
+-   Run through some sample workflows. Just a few for example:
+    -   Adding a node
+    -   Adding a ClusterCIDRConfig
+    -   Deleting a ClusterCIDRConfig that is in use
```

Comment:
I think the user stories describe good e2e tests, we should have e2e for each user story.

Reply:
Acknowledged. Adding.